### ISSUE TITLE
Notification plugin stage status events don't include accurate material information for fields such as git shallow-clone

### ISSUE THREAD

**Issue by maber1993** [gocd/gocd#12574]
##### Issue Type
- Bug Report

##### Summary
During the debugging process using the official example, I noticed that there is a discrepancy between the shallowclone value in the request body and the panel configuration result after the stage status event is triggered.

###### Basic environment details

* Go Version: `gocd/gocd-server:v23.5.0`
* JAVA Version: `17`
* OS: `ubuntu 23.10`
* plugin：`https://github.com/gocd-contrib/notification-skeleton-plugin`

##### Steps to Reproduce

1. `cp /data/workspace/notification-skeleton-plugin-master/build/libs/notification-skeleton-plugin-0.1-SNAPSHOT.jar /data/gocd/plugins/external/`
2. `docker run -it --rm -p 8153:8153 -v /data/gocd:/godata gocd/gocd-server:v23.5.0`
3. Triggered pipeline

##### Actual Results
The correct configuration values are obtained in the notification plugin.

##### Log snippets

```
jvm 1    | 2024-03-05 07:17:41,711  INFO [180@MessageListener for PluginNotificationMessageListener] p.c.e.n.s.c.e.n.ExamplePlugin:72 [plugin-com.example.notification.systemout] - {"pipeline":{"name":"test","counter":"1","label":"1","group":"test2","build-cause":[{"material":{"git-configuration":{"shallow-clone":false,"branch":"master","url":"https://*************/test.git"},"type":"git"},"changed":true,"modifications":[{"revision":"679ee42eb627758ac898bb1ef450636abfbfa27e","modified-time":"2024-03-04T07:30:02.000+0000","data":{}}]}],"stage":{"name":"build","counter":"1","approval-type":"success","approved-by":"changes","state":"Failed","result":"Failed","create-time":"2024-03-05T07:17:13.804+0000","last-transition-time":"2024-03-05T07:17:41.705+0000","jobs":[{"name":"build","schedule-time":"2024-03-05T07:17:13.804+0000","assign-time":"2024-03-05T07:17:25.029+0000","complete-time":"2024-03-05T07:17:38.365+0000","state":"Completed","result":"Failed","agent-uuid":"ce1c2508-2373-44fa-9b4b-fdfbe3b7d2c2"},{"name":"build2","schedule-time":"2024-03-05T07:17:13.804+0000","assign-time":"2024-03-05T07:17:39.383+0000","complete-time":"2024-03-05T07:17:41.705+0000","state":"Completed","result":"Passed","agent-uuid":"ce1c2508-2373-44fa-9b4b-fdfbe3b7d2c2"}]}}}

```
-->

##### Code snippets/Screenshots

![image](https://github.com/gocd/gocd/assets/18278619/725904e4-dfb9-4b5b-9960-636dffb8fb13)
![image](https://github.com/gocd/gocd/assets/18278619/5b69b97c-1040-4d3f-a5eb-56dcce771822)
![image](https://github.com/gocd/gocd/assets/18278619/fa255d3b-2958-4880-83cc-3fdb05a91aaa)
-->

**Comment by chadlwilson** [comment]
Yeah, these notifications appear to be based on data captured in the database for an individual pipeline run; and I suspect the material specific values such as `shallow-clone` are not captured for every individual stage run, yet still misleadingly included in the serialized configuration.

Do you have a need specifically for this field in a custom plugin you are writing, or just something weird you noticed?

**Comment by maber1993** [comment]
> Yeah, these notifications appear to be based on data captured in the database for an individual pipeline run; and I suspect the material specific values such as `shallow-clone` are not captured for every individual stage run, yet still misleadingly included in the serialized configuration.是的，这些通知似乎是基于数据库中捕获的单个管道运行的数据；我怀疑诸如 `shallow-clone` 之类的材料特定值并未针对每个单独的阶段运行捕获，但仍然误导性地包含在序列化配置中。
> 
> Do you have a need specifically for this field in a custom plugin you are writing, or just something weird you noticed?您是否在您正在编写的自定义插件中专门需要此字段，或者只是您注意到一些奇怪的东西？

Yes, I am currently trying to develop a notification plugin. To expedite the build process, I am paying attention to the "shallowclone".

**Comment by chadlwilson** [comment]
> Yes, I am currently trying to develop a notification plugin. To expedite the build process, I am paying attention to the `shallow-clone`.

For validating configuration data like this, it's possibly better to achieve this by querying the [GoCD APIs for materials](https://api.gocd.org/current/#materials), rather than relying on the notification data when every stage status changes.

**Comment by maber1993** [comment]
> > Yes, I am currently trying to develop a notification plugin. To expedite the build process, I am paying attention to the `shallow-clone`.
> 
> For validating configuration data like this, it's possibly better to achieve this by querying the [GoCD APIs for materials](https://api.gocd.org/current/#materials), rather than relying on the notification data when every stage status changes.

Yes, the API can fetch the correct status. Initially, the purpose of developing the plugin was to promptly detect when a stage fails to run. During the development process, I came across the shallow-clone parameter and planned to merge our initial detection functionality. However, it seems that this approach is not feasible at the moment.

> > Yeah, these notifications appear to be based on data captured in the database for an individual pipeline run; and I suspect the material specific values such as `shallow-clone` are not captured for every individual stage run, yet still misleadingly included in the serialized configuration.
> > Do you have a need specifically for this field in a custom plugin you are writing, or just something weird you noticed?
> 
> Yes, I am currently trying to develop a notification plugin. To expedite the build process, I am paying attention to the `shallow-clone`.

**Comment by maber1993** [comment]
> > Yes, I am currently trying to develop a notification plugin. To expedite the build process, I am paying attention to the `shallow-clone`.
> 
> For validating configuration data like this, it's possibly better to achieve this by querying the [GoCD APIs for materials](https://api.gocd.org/current/#materials), rather than relying on the notification data when every stage status changes.

Yes, the API can fetch the correct status. Initially, the purpose of developing the plugin was to promptly detect when a stage fails to run. During the development process, I came across the shallow-clone parameter and planned to merge our initial detection functionality. However, it seems that this approach is not feasible at the moment.
